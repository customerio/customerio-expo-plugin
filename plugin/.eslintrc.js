const config = require("expo-module-scripts/eslintrc.base.js");

if (!config.rules) config.rules = {};
if (!config.extends) config.extends = [];
if (!config.plugins) config.plugins = [];

config.rules["prettier/prettier"] = "error";
config.extends.push("prettier");
config.plugins.push("prettier");

module.exports = config;
