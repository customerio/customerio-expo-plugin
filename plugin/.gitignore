build
lib

# Node / Yarn
node_modules
*.log
.nvm
npm-debug.log
yarn-error.log

# OS X
.DS_Store

# Emacs
*~

#jest
coverage/

# VSCode
.vscode/
