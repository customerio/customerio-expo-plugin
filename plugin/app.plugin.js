// For development testing, we directly require the source files
// This allows us to test changes without needing to rebuild the plugin
try {
  module.exports = require("./lib/commonjs/index");
} catch (error) {
  // Fallback to src for development
  console.log("Using source files directly for testing - attempting to register ts-node");
  try {
    // Try to register ts-node for TypeScript compilation on the fly
    require('ts-node').register({
      transpileOnly: true,
      compilerOptions: {
        module: "commonjs",
        target: "es2017",
      }
    });
    module.exports = require("./src/index");
  } catch (tsNodeError) {
    console.error("Could not load ts-node for TypeScript compilation:", tsNodeError.message);
    console.log("Trying to load compiled JavaScript files if they exist...");
    
    // As a last resort, try to find any .js files that may exist
    try {
      const fs = require('fs');
      const path = require('path');
      
      // Check if index.js exists in the src directory (might have been compiled but not moved)
      const indexJsPath = path.resolve(__dirname, 'src', 'index.js');
      if (fs.existsSync(indexJsPath)) {
        console.log(`Found ${indexJsPath}, attempting to load it`);
        module.exports = require(indexJsPath);
      } else {
        console.error("Could not find any suitable module to load. Please rebuild the plugin.");
        throw new Error("CustomerIO Expo Plugin failed to load. Please rebuild the plugin with 'npm run build'.");
      }
    } catch (finalError) {
      console.error("Fatal error loading CustomerIO Expo Plugin:", finalError);
      throw finalError;
    }
  }
}
