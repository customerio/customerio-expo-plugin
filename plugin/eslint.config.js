// @ts-check

const eslint = require('@eslint/js');
const tseslint = require('typescript-eslint');

module.exports = [
  {
    ignores: ['lib/**', 'node_modules/**', '**/*.js', '**/*.d.ts', 'eslint.config.js'],
  },
  eslint.configs.recommended,
  ...tseslint.configs.recommended.map(config => ({
    ...config,
    rules: {
      ...config.rules,
      '@typescript-eslint/no-require-imports': 'off',
    }
  })),
  {
    files: ['plugin/src/**/*.ts'],
    languageOptions: {
      parser: tseslint.parser,
      parserOptions: {
        project: './tsconfig.json',
        tsconfigRootDir: __dirname + '/..',
      },
    },
    plugins: {
      '@typescript-eslint': tseslint.plugin,
    },
    rules: {
      // Allow require() imports for Node.js compatibility
      '@typescript-eslint/no-var-requires': 'off',
      '@typescript-eslint/no-require-imports': 'off',
      
      // Keep these as warnings to gradually improve
      '@typescript-eslint/no-explicit-any': 'warn',
      '@typescript-eslint/no-non-null-assertion': 'warn',
      
      // Ignore unused vars with underscore prefix
      '@typescript-eslint/no-unused-vars': ['error', { 
        argsIgnorePattern: '^_',
        varsIgnorePattern: '^_',
        caughtErrorsIgnorePattern: '^_'
      }],
    },
  }
];
