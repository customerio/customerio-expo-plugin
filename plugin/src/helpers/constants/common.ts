/**
 * Platform constants for native SDK initialization
 */
export const PLATFORM = {
  IOS: 'ios',
  ANDROID: 'android',
} as const;

/**
 * Platform type definition
 */
export type Platform = typeof PLATFORM[keyof typeof PLATFORM];
