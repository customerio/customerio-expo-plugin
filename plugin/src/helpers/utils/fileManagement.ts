import {
  readFile,
  writeFile,
  appendFile,
  existsSync,
  copyFileSync,
  mkdirSync,
  writeFileSync,
  readFileSync,
  MakeDirectoryOptions,
} from 'fs';

export class FileManagement {
  static async read(path: string): Promise<string> {
    return new Promise<string>((resolve, reject) => {
      readFile(path, 'utf8', (err, data) => {
        if (err || !data) {
          reject(err);
          return;
        }
        resolve(data);
      });
    });
  }

  static async write(path: string, contents: string): Promise<void> {
    return new Promise<void>((resolve, reject) => {
      writeFile(path, contents, 'utf8', (err) => {
        if (err) {
          reject(err);
          return;
        }
        resolve();
      });
    });
  }

  static async append(path: string, contents: string): Promise<void> {
    return new Promise<void>((resolve, reject) => {
      appendFile(path, contents, 'utf8', (err) => {
        if (err) {
          reject(err);
          return;
        }
        resolve();
      });
    });
  }

  static exists(path: string) {
    return existsSync(path);
  }

  static copyFile(src: string, dest: string) {
    try {
      copyFileSync(src, dest);
    } catch (err) {
      console.log(`Error copying file from ${src} to ${dest}: `, err);
    }
  }

  static mkdir(path: string, options: MakeDirectoryOptions) {
    try {
      mkdirSync(path, options);
    } catch (err) {
      console.log(`Error creating directory ${path}: `, err);
    }
  }

  static writeFile(path: string, data: string) {
    try {
      writeFileSync(path, data);
    } catch (err) {
      console.log(`Error writing to file ${path}: `, err);
    }
  }

  static readFile(path: string) {
    try {
      return readFileSync(path, 'utf-8');
    } catch (err) {
      console.log(`Error reading file ${path}: `, err);
    }

    return '';
  }
}
