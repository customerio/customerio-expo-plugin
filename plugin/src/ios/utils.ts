import type { CustomerIOPluginOptionsIOS } from '../types/cio-types';

/**
 * Returns t
 * @param iosOptions The plugin iOS configuration options
 * @returns true if FCM is configured to be used as push provider
 */
export const isFcmPushProvider = (iosOptions?: CustomerIOPluginOptionsIOS): boolean => {
    return iosOptions?.pushNotification?.provider === "fcm";
};