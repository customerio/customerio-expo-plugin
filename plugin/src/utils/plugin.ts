import fs from 'fs';
import path from 'path';

const findPluginPackageRoot = (): string => {
  const finder = require('find-package-json');
  const f = finder(__dirname);
  const root = f.next().filename;
  return path.dirname(root);
};

const pluginPackageRoot = findPluginPackageRoot();

// Returns path to plugin's native template files directory
export const getNativeFilesPath = (): string => {
  return path.join(
    pluginPackageRoot,
    'plugin/src/helpers/native-files/'
  );
};

// Returns path to plugin's Android native template files
export const getAndroidNativeFilesPath = (): string => {
  return path.join(getNativeFilesPath(), 'android');
};

// Returns path to plugin's iOS native template files
export const getIosNativeFilesPath = (): string => {
  return path.join(getNativeFilesPath(), 'ios');
};

// Reads the version of the plugin from its `package.json` and returns it as a string.
export const getPluginVersion = (): string => {
  const packageJsonPath = path.resolve(
    pluginPackageRoot,
    'package.json'
  );
  if (!fs.existsSync(packageJsonPath)) {
    throw new Error(`package.json not found at ${packageJsonPath}`);
  }

  const packageJson = JSON.parse(fs.readFileSync(packageJsonPath, 'utf8'));
  if (!packageJson.version) {
    throw new Error(`"version" field is missing in ${packageJsonPath}`);
  }
  return packageJson.version;
};
