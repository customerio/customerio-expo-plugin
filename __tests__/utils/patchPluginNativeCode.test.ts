import path from 'path';
import { PLATFORM } from '../../plugin/src/helpers/constants/common';
import { FileManagement } from '../../plugin/src/helpers/utils/fileManagement';
import {
  patchNativeSDKInitializer,
} from '../../plugin/src/helpers/utils/patchPluginNativeCode';
import type { NativeSDKConfig } from '../../plugin/src/types/cio-types';
import { getPluginVersion } from '../../plugin/src/utils/plugin';

// Mock dependencies
jest.mock('../../plugin/src/utils/plugin');
const mockGetPluginVersion = getPluginVersion as jest.MockedFunction<typeof getPluginVersion>;

describe('Native SDK Configuration Patching', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    mockGetPluginVersion.mockReturnValue('2.5.0');
  });

  describe('patchNativeSDKInitializer()', () => {
    const baseSdkConfig: NativeSDKConfig = {
      cdpApiKey: 'test-api-key',
    };

    const mockContent = `
      version: {{EXPO_PLUGIN_VERSION}}
      apiKey: {{CDP_API_KEY}}
      region: {{REGION}}
      logLevel: {{LOG_LEVEL}}
      autoTrack: {{AUTO_TRACK_DEVICE_ATTRIBUTES}}
      lifecycle: {{TRACK_APPLICATION_LIFECYCLE_EVENTS}}
      screenView: {{SCREEN_VIEW_USE}}
      siteId: {{SITE_ID}}
      migrationSiteId: {{MIGRATION_SITE_ID}}
    `;

    // Read the actual template files
    const swiftTemplateContent = FileManagement.readFile(
      path.join(__dirname, '../../plugin/src/helpers/native-files/ios/CustomerIOSDKInitializer.swift')
    );

    const kotlinTemplateContent = FileManagement.readFile(
      path.join(__dirname, '../../plugin/src/helpers/native-files/android/CustomerIOSDKInitializer.kt')
    );

    test('requires platform parameter for TypeScript compilation', () => {
      // This test ensures the platform parameter is required
      expect(() => {
        // @ts-expect-error Testing that platform parameter is required
        patchNativeSDKInitializer(mockContent, baseSdkConfig);
      }).toBeDefined(); // Just testing TypeScript compilation requirement
    });

    test('replaces undefined optional values with nil fallbacks on iOS platform', () => {
      const result = patchNativeSDKInitializer(mockContent, PLATFORM.IOS, baseSdkConfig);

      expect(result).toContain('version: 2.5.0');
      expect(result).toContain('apiKey: test-api-key');
      expect(result).toContain('region: ""'); // Empty string fallback for region (nil not supported)
      expect(result).toContain('logLevel: nil');
      expect(result).toContain('autoTrack: nil');
      expect(result).toContain('lifecycle: nil');
      expect(result).toContain('screenView: nil');
    });

    test('replaces undefined optional values with null fallbacks on Android platform', () => {
      const result = patchNativeSDKInitializer(mockContent, PLATFORM.ANDROID, baseSdkConfig);

      expect(result).toContain('version: 2.5.0');
      expect(result).toContain('apiKey: test-api-key');
      expect(result).toContain('region: ""'); // Empty string fallback for region (nil not supported)
      expect(result).toContain('logLevel: null'); // null for Android platform
      expect(result).toContain('autoTrack: null');
      expect(result).toContain('lifecycle: null');
      expect(result).toContain('screenView: null');
    });


    test('preserves explicit boolean values for both true and false', () => {
      const testCases = [true, false];

      testCases.forEach(value => {
        const config = {
          ...baseSdkConfig,
          autoTrackDeviceAttributes: value,
          trackApplicationLifecycleEvents: value,
        };
        const iosResult = patchNativeSDKInitializer(mockContent, PLATFORM.IOS, config);
        const androidResult = patchNativeSDKInitializer(mockContent, PLATFORM.ANDROID, config);

        expect(iosResult).toContain(`autoTrack: ${value}`);
        expect(iosResult).toContain(`lifecycle: ${value}`);
        expect(androidResult).toContain(`autoTrack: ${value}`);
        expect(androidResult).toContain(`lifecycle: ${value}`);
      });
    });


    test('handles all valid screenViewUse enum values on both platforms', () => {
      const testCases = ['all', 'inapp'] as const;

      testCases.forEach(screenViewUse => {
        const config = { ...baseSdkConfig, screenViewUse };
        const iosResult = patchNativeSDKInitializer(mockContent, PLATFORM.IOS, config);
        const androidResult = patchNativeSDKInitializer(mockContent, PLATFORM.ANDROID, config);

        expect(iosResult).toContain(`screenView: "${screenViewUse}"`);
        expect(androidResult).toContain(`screenView: "${screenViewUse}"`);
      });
    });

    test('handles all valid logLevel enum values on both platforms', () => {
      const testCases = ['debug', 'info', 'error', 'none'] as const;

      testCases.forEach(logLevel => {
        const config = { ...baseSdkConfig, logLevel };
        const iosResult = patchNativeSDKInitializer(mockContent, PLATFORM.IOS, config);
        const androidResult = patchNativeSDKInitializer(mockContent, PLATFORM.ANDROID, config);

        expect(iosResult).toContain(`logLevel: "${logLevel}"`);
        expect(androidResult).toContain(`logLevel: "${logLevel}"`);
      });
    });

    test('handles all valid region enum values on both platforms', () => {
      const testCases = ['US', 'EU'] as const;

      testCases.forEach(region => {
        const config = { ...baseSdkConfig, region };
        const iosResult = patchNativeSDKInitializer(mockContent, PLATFORM.IOS, config);
        const androidResult = patchNativeSDKInitializer(mockContent, PLATFORM.ANDROID, config);

        expect(iosResult).toContain(`region: "${region}"`);
        expect(androidResult).toContain(`region: "${region}"`);
      });
    });


    describe('siteId and migrationSiteId business logic', () => {
      test('uses explicit migrationSiteId when provided', () => {
        const config = { ...baseSdkConfig, migrationSiteId: 'explicit-migration-id' };
        const iosResult = patchNativeSDKInitializer(mockContent, PLATFORM.IOS, config);
        const androidResult = patchNativeSDKInitializer(mockContent, PLATFORM.ANDROID, config);

        // migrationSiteId = migrationSiteId (when defined)
        expect(iosResult).toContain('migrationSiteId: "explicit-migration-id"');
        expect(androidResult).toContain('migrationSiteId: "explicit-migration-id"');
        // siteId should be nil/null when only migrationSiteId provided
        expect(iosResult).toContain('siteId: nil');
        expect(androidResult).toContain('siteId: null');
      });

      test('copies siteId to migrationSiteId when migrationSiteId is undefined', () => {
        const config = { ...baseSdkConfig, siteId: 'main-site-id' };
        const iosResult = patchNativeSDKInitializer(mockContent, PLATFORM.IOS, config);
        const androidResult = patchNativeSDKInitializer(mockContent, PLATFORM.ANDROID, config);

        // migrationSiteId = siteId (when migrationSiteId undefined and siteId defined)
        expect(iosResult).toContain('migrationSiteId: "main-site-id"');
        expect(androidResult).toContain('migrationSiteId: "main-site-id"');
        expect(iosResult).toContain('siteId: "main-site-id"');
        expect(androidResult).toContain('siteId: "main-site-id"');
      });

      test('sets both siteId and migrationSiteId to platform null when both are undefined', () => {
        const iosResult = patchNativeSDKInitializer(mockContent, PLATFORM.IOS, baseSdkConfig);
        const androidResult = patchNativeSDKInitializer(mockContent, PLATFORM.ANDROID, baseSdkConfig);

        // migrationSiteId = nil/null (when both undefined)
        expect(iosResult).toContain('migrationSiteId: nil');
        expect(androidResult).toContain('migrationSiteId: null');
        expect(iosResult).toContain('siteId: nil');
        expect(androidResult).toContain('siteId: null');
      });

      test('preserves distinct values when both siteId and migrationSiteId are provided', () => {
        const config = { ...baseSdkConfig, siteId: 'site123', migrationSiteId: 'migration456' };
        const iosResult = patchNativeSDKInitializer(mockContent, PLATFORM.IOS, config);
        const androidResult = patchNativeSDKInitializer(mockContent, PLATFORM.ANDROID, config);

        expect(iosResult).toContain('siteId: "site123"');
        expect(iosResult).toContain('migrationSiteId: "migration456"');
        expect(androidResult).toContain('siteId: "site123"');
        expect(androidResult).toContain('migrationSiteId: "migration456"');
      });

      test('treats empty or whitespace-only siteId as undefined and uses platform null', () => {
        const testCases = ['', '   '];

        testCases.forEach(siteId => {
          const config = { ...baseSdkConfig, siteId };
          const iosResult = patchNativeSDKInitializer(mockContent, PLATFORM.IOS, config);
          const androidResult = patchNativeSDKInitializer(mockContent, PLATFORM.ANDROID, config);

          expect(iosResult).toContain('siteId: nil');
          expect(iosResult).toContain('migrationSiteId: nil');
          expect(androidResult).toContain('siteId: null');
          expect(androidResult).toContain('migrationSiteId: null');
        });
      });
    });

    describe('snapshots', () => {
      test('iOS complete', () => {
        const config = {
          ...baseSdkConfig,
          logLevel: 'debug' as const,
          region: 'US' as const,
          siteId: 'test-site-123',
          autoTrackDeviceAttributes: true,
          trackApplicationLifecycleEvents: false,
          screenViewUse: 'inapp' as const,
        };

        const result = patchNativeSDKInitializer(swiftTemplateContent, PLATFORM.IOS, config);
        expect(result).toMatchSnapshot();
      });

      test('iOS minimal', () => {
        const result = patchNativeSDKInitializer(swiftTemplateContent, PLATFORM.IOS, baseSdkConfig);
        expect(result).toMatchSnapshot();
      });

      test('Android complete', () => {
        const config = {
          ...baseSdkConfig,
          logLevel: 'info' as const,
          region: 'EU' as const,
          siteId: 'android-site-456',
          autoTrackDeviceAttributes: false,
          trackApplicationLifecycleEvents: true,
          screenViewUse: 'all' as const,
        };

        const result = patchNativeSDKInitializer(kotlinTemplateContent, PLATFORM.ANDROID, config);
        expect(result).toMatchSnapshot();
      });

      test('Android minimal', () => {
        const result = patchNativeSDKInitializer(kotlinTemplateContent, PLATFORM.ANDROID, baseSdkConfig);
        expect(result).toMatchSnapshot();
      });
    });
  });
});
