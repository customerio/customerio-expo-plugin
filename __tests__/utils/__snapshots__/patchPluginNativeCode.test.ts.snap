// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Native SDK Configuration Patching patchNativeSDKInitializer() snapshots iOS complete 1`] = `
"import CioDataPipelines
import CioInternalCommon
import CioMessagingInApp

class CustomerIOSDKInitializer {
    static func initialize() {
        // Override SDK client info to include Expo metadata in user agent
        let pluginVersion = "2.5.0"
        DIGraphShared.shared.override(
            value: CustomerIOSdkClient(source: "Expo", sdkVersion: pluginVersion),
            forType: SdkClient.self
        )

        let cdpApiKey = "test-api-key"
        let siteId: String? = "test-site-123"
        let region = CioInternalCommon.Region.getRegion(from: "US")

        let builder = SDKConfigBuilder(cdpApiKey: cdpApiKey)
        setIfDefined(value: "debug", thenPassItTo: builder.logLevel, transformingBy: CioLogLevel.getLogLevel)
        setIfDefined(value: region, thenPassItTo: builder.region)
        setIfDefined(value: true, thenPassItTo: builder.autoTrackDeviceAttributes)
        setIfDefined(value: false, thenPassItTo: builder.trackApplicationLifecycleEvents)
        setIfDefined(value: "inapp", thenPassItTo: builder.screenViewUse) { ScreenView.getScreenView($0) }
        setIfDefined(value: "test-site-123", thenPassItTo: builder.migrationSiteId)

        CustomerIO.initialize(withConfig: builder.build())

        if let siteId = siteId {
            let inAppConfig = MessagingInAppConfigBuilder(siteId: siteId, region: region).build()
            MessagingInApp.initialize(withConfig: inAppConfig)
            // Retrieves ReactInAppEventListener from DI graph, populated when it is accessed in React Native SDK.
            if let listener: InAppEventListener? = DIGraphShared.shared.getOverriddenInstance() {
                MessagingInApp.shared.setEventListener(listener)
            }
        }
    }

    /// Apply a value to a setter only if it's non-nil
    private static func setIfDefined<Raw>(
        value rawValue: Raw?,
        thenPassItTo handler: (Raw) -> Any
    ) {
        setIfDefined(value: rawValue, thenPassItTo: handler) { $0 }
    }

    /// Apply a value after transforming it, only if both the original and transformed values are non-nil
    private static func setIfDefined<Raw, Transformed>(
        value rawValue: Raw?,
        thenPassItTo handler: (Transformed) -> Any,
        transformingBy transform: (Raw) -> Transformed?
    ) {
        if let value = rawValue, let result = transform(value) {
            _ = handler(result)
        }
    }
}
"
`;

exports[`Native SDK Configuration Patching patchNativeSDKInitializer() snapshots iOS minimal 1`] = `
"import CioDataPipelines
import CioInternalCommon
import CioMessagingInApp

class CustomerIOSDKInitializer {
    static func initialize() {
        // Override SDK client info to include Expo metadata in user agent
        let pluginVersion = "2.5.0"
        DIGraphShared.shared.override(
            value: CustomerIOSdkClient(source: "Expo", sdkVersion: pluginVersion),
            forType: SdkClient.self
        )

        let cdpApiKey = "test-api-key"
        let siteId: String? = nil
        let region = CioInternalCommon.Region.getRegion(from: "")

        let builder = SDKConfigBuilder(cdpApiKey: cdpApiKey)
        setIfDefined(value: nil, thenPassItTo: builder.logLevel, transformingBy: CioLogLevel.getLogLevel)
        setIfDefined(value: region, thenPassItTo: builder.region)
        setIfDefined(value: nil, thenPassItTo: builder.autoTrackDeviceAttributes)
        setIfDefined(value: nil, thenPassItTo: builder.trackApplicationLifecycleEvents)
        setIfDefined(value: nil, thenPassItTo: builder.screenViewUse) { ScreenView.getScreenView($0) }
        setIfDefined(value: nil, thenPassItTo: builder.migrationSiteId)

        CustomerIO.initialize(withConfig: builder.build())

        if let siteId = siteId {
            let inAppConfig = MessagingInAppConfigBuilder(siteId: siteId, region: region).build()
            MessagingInApp.initialize(withConfig: inAppConfig)
            // Retrieves ReactInAppEventListener from DI graph, populated when it is accessed in React Native SDK.
            if let listener: InAppEventListener? = DIGraphShared.shared.getOverriddenInstance() {
                MessagingInApp.shared.setEventListener(listener)
            }
        }
    }

    /// Apply a value to a setter only if it's non-nil
    private static func setIfDefined<Raw>(
        value rawValue: Raw?,
        thenPassItTo handler: (Raw) -> Any
    ) {
        setIfDefined(value: rawValue, thenPassItTo: handler) { $0 }
    }

    /// Apply a value after transforming it, only if both the original and transformed values are non-nil
    private static func setIfDefined<Raw, Transformed>(
        value rawValue: Raw?,
        thenPassItTo handler: (Transformed) -> Any,
        transformingBy transform: (Raw) -> Transformed?
    ) {
        if let value = rawValue, let result = transform(value) {
            _ = handler(result)
        }
    }
}
"
`;

exports[`Native SDK Configuration Patching patchNativeSDKInitializer() snapshots Android complete 1`] = `
"package io.customer.sdk.expo

import android.app.Application
import io.customer.datapipelines.config.ScreenView
import io.customer.messaginginapp.MessagingInAppModuleConfig
import io.customer.messaginginapp.ModuleMessagingInApp
import io.customer.messagingpush.MessagingPushModuleConfig
import io.customer.messagingpush.ModuleMessagingPushFCM
import io.customer.reactnative.sdk.messaginginapp.ReactInAppEventListener
import io.customer.sdk.CustomerIOBuilder
import io.customer.sdk.core.util.CioLogLevel
import io.customer.sdk.data.model.Region

object CustomerIOSDKInitializer {
    fun initialize(application: Application) = with(
        CustomerIOBuilder(application, "test-api-key")
    ) {
        val siteId: String? = "android-site-456"
        val migrationSiteId: String? = "android-site-456"
        val region = Region.getRegion("EU")

        setIfDefined("info", CustomerIOBuilder::logLevel) { CioLogLevel.getLogLevel(it) }
        setIfDefined(region, CustomerIOBuilder::region)
        setIfDefined(false, CustomerIOBuilder::autoTrackDeviceAttributes)
        setIfDefined(true, CustomerIOBuilder::trackApplicationLifecycleEvents)
        setIfDefined("all", CustomerIOBuilder::screenViewUse) { ScreenView.getScreenView(it) }
        setIfDefined(migrationSiteId, CustomerIOBuilder::migrationSiteId)

        // Add messaging modules if siteId is provided
        if (!(siteId.isNullOrBlank())) {
            addCustomerIOModule(
                ModuleMessagingInApp(
                    MessagingInAppModuleConfig.Builder(siteId, region)
                        .setEventListener(ReactInAppEventListener())
                        .build()
                )
            )
        }
        addCustomerIOModule(
            ModuleMessagingPushFCM(
                MessagingPushModuleConfig.Builder().build()
            )
        )

        build()
    }
}

// Apply a value after transforming it, only if both the original and transformed values are non-nil
private inline fun <R, T> CustomerIOBuilder.setIfDefined(
    value: R?,
    block: CustomerIOBuilder.(T) -> CustomerIOBuilder,
    transform: (R) -> T,
): CustomerIOBuilder = value?.let { block(transform(it)) } ?: this

// Apply a value to a setter only if it's non-nil
private inline fun <T> CustomerIOBuilder.setIfDefined(
    value: T?,
    block: CustomerIOBuilder.(T) -> CustomerIOBuilder,
): CustomerIOBuilder = setIfDefined(
    value = value,
    block = block,
    transform = { it },
)
"
`;

exports[`Native SDK Configuration Patching patchNativeSDKInitializer() snapshots Android minimal 1`] = `
"package io.customer.sdk.expo

import android.app.Application
import io.customer.datapipelines.config.ScreenView
import io.customer.messaginginapp.MessagingInAppModuleConfig
import io.customer.messaginginapp.ModuleMessagingInApp
import io.customer.messagingpush.MessagingPushModuleConfig
import io.customer.messagingpush.ModuleMessagingPushFCM
import io.customer.reactnative.sdk.messaginginapp.ReactInAppEventListener
import io.customer.sdk.CustomerIOBuilder
import io.customer.sdk.core.util.CioLogLevel
import io.customer.sdk.data.model.Region

object CustomerIOSDKInitializer {
    fun initialize(application: Application) = with(
        CustomerIOBuilder(application, "test-api-key")
    ) {
        val siteId: String? = null
        val migrationSiteId: String? = null
        val region = Region.getRegion("")

        setIfDefined(null, CustomerIOBuilder::logLevel) { CioLogLevel.getLogLevel(it) }
        setIfDefined(region, CustomerIOBuilder::region)
        setIfDefined(null, CustomerIOBuilder::autoTrackDeviceAttributes)
        setIfDefined(null, CustomerIOBuilder::trackApplicationLifecycleEvents)
        setIfDefined(null, CustomerIOBuilder::screenViewUse) { ScreenView.getScreenView(it) }
        setIfDefined(migrationSiteId, CustomerIOBuilder::migrationSiteId)

        // Add messaging modules if siteId is provided
        if (!(siteId.isNullOrBlank())) {
            addCustomerIOModule(
                ModuleMessagingInApp(
                    MessagingInAppModuleConfig.Builder(siteId, region)
                        .setEventListener(ReactInAppEventListener())
                        .build()
                )
            )
        }
        addCustomerIOModule(
            ModuleMessagingPushFCM(
                MessagingPushModuleConfig.Builder().build()
            )
        )

        build()
    }
}

// Apply a value after transforming it, only if both the original and transformed values are non-nil
private inline fun <R, T> CustomerIOBuilder.setIfDefined(
    value: R?,
    block: CustomerIOBuilder.(T) -> CustomerIOBuilder,
    transform: (R) -> T,
): CustomerIOBuilder = value?.let { block(transform(it)) } ?: this

// Apply a value to a setter only if it's non-nil
private inline fun <T> CustomerIOBuilder.setIfDefined(
    value: T?,
    block: CustomerIOBuilder.(T) -> CustomerIOBuilder,
): CustomerIOBuilder = setIfDefined(
    value = value,
    block = block,
    transform = { it },
)
"
`;
